data/collector.go

```diff
@@ -307,11 +308,34 @@ func FailedToReadErr(component, method string, err error) error {
 	return errors.Errorf("failed to get reading of method %s of component %s: %v", method, component, err)
 }
 
-// GetExtraFromContext sets the extra struct with "fromDataManagement": true if the flag is true in the context.
-func GetExtraFromContext(ctx context.Context) (*structpb.Struct, error) {
-	extra := make(map[string]interface{})
+// GetExtraFromContext adds "fromDataManagement": true to the extra map if the flag is true in the context, and returns a protobuf Struct.
+func GetExtraFromContext(ctx context.Context, extra map[string]interface{}) (*structpb.Struct, error) {
 	if ctx.Value(FromDMContextKey{}) == true {
 		extra[FromDMString] = true
 	}
 	return protoutils.StructToStructPb(extra)
 }
+
+type requestWithExtra interface {
+	GetExtra() *structpb.Struct
+}
+
+// UnaryClientInterceptor adds "fromDataManagement": true to the outgoing req's Extra struct if the flag is true in the context.
+func UnaryClientInterceptor(
```

Comment:
Yes, I tested this by excluding the camera's client changes. If we do decide to go with the interceptor approach we won't need the client changes that are currently in this PR (tried to point this out in [this](https://github.com/viamrobotics/rdk/pull/2830#discussion_r1304866505) comment) - I left them in here to get feedback on which approach was preferred from reviewers, but maybe that's more confusing than helpful 😅 

> Would this also work with a custom component?

Yes, tested with custom model; the interceptor here is on the module manager (as opposed to being set up on the robot client [here](https://github.com/viamrobotics/rdk/blob/04e5458f7bc46bb7233fe5f5ef5aa12001de5ede/robot/client/client.go#L280)).

Reply:
Oh got it, I think I missed the comment. I am in favor of the interceptor approach! 

Thanks for testing. :)

Reply:
Actually thinking on this a bit more - this works with custom models, but I am less confident about custom types. @Otterverse do you know if requests for a modular resource that implements a custom type would follow the same path through the module manager interceptors?

Reply:
As noted in the scope document, using (grpc) metadata was rejected for some good reasons, and this feels like it's trying to skirt around that by using metadata (contexts) for part of things anyway, but it runs into similar issues. That is, it only works for APIs (resource types) that are built in, and you can modify the client package for.

While the interceptor is a clever approach, I worry it makes the code harder to understand, as it's setting a contextual value, and then semi-hidden "magic" happens at the module layer to add the actual proto-defined fields to the request.

My personal feeling is that directly just adding the "extra" field/value at the time the request is made would be much more readable and maintainable in the long run.

If I'm missing something on why that would be more difficult or more complex, please let me know. But as I see it, it's basically taking:

```
ctx = context.WithValue(ctx, data.FromDMContextKey{}, true)
v, err := arm.EndPosition(ctx, nil)
```
And turning it into
```
v, err := arm.EndPosition(ctx, data.FromDMExtraMap)
```
(Assuming data.FromDMExtraMap is defined as a map that contains the FromDM boolean already set)

That seems simpler and much more straight forward to me, and also provides an example modular component authors making custom types can directly copy as well.

Reply:
Thanks for the feedback James! The approach you detailed would work for most of our builtin types, and agree that this is more direct and clearer to read - I can make these changes in the next commit.

The camera requests are an edge case to this since the underlying functions don't take extra (adding extra to [ReadImage](https://github.com/viamrobotics/rdk/blob/1b1bc1f590229a2c065d7785280e79f02b23ab47/components/camera/camera.go#L99) is out of scope for this project due to the eng complexity of threading it through the gostream api, but I can check with AV to see if adding extra to NextPointCloud would be feasible). The client interceptor approach doesn't seem necessary for just handling this one case, so removing that and adding back in the client code to add the extra in the camera request based on metadata.

Reply:
Thank you James! It was my idea to add the interceptor. The main advantage would be that new component types would get this for free without having to add it explicitly in the code, but maybe it's better to make it explicit. Appreciate you weighing in and explaining your thought process!